File: cpp/139_word_break.cpp
#include <iostream>
#include <vector>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <string>
#include <ctime>
#include "utils.h"
using namespace std;

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<bool> mem(s.size()+1, false);
        mem[0] = true;

        for (int i = 1; i <= s.size(); i++) {
            for (int j = i-1; j >= 0; j--) {
                if (mem[j]) {
                    string tmp = s.substr(j, i-j);
                    if (find(wordDict.begin(), wordDict.end(), tmp) != wordDict.end()) {
                        mem[i] = true;
                        break;
                    }
                }
            }
        }

        return mem[s.size()];

    }


};

int main() {

    string s = "leetcode";
    vector<string> wordDict = {"leet", "code"};
    cout << Solution().wordBreak(s, wordDict) << endl;

    s = "bb";
    wordDict = {"a", "b", "bbb", "bbbb"};
    cout << Solution().wordBreak(s, wordDict) << endl;

    s = "leettcode";
    wordDict = {"leet", "code"};
    cout << Solution().wordBreak(s, wordDict) << endl;

    s = "leetcode";
    wordDict = {"lee", "code"};
    cout << Solution().wordBreak(s, wordDict) << endl;

    s = "leetcde";
    wordDict = {"leet", "code"};
    cout << Solution().wordBreak(s, wordDict) << endl;

    return 0;
}

File: leetcode_py/139_word_break.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        wordSet = set(wordDict)
        mem = [False] * (len(s)+1)
        mem[0] = True
        for i in range(len(s)):
            for j in range(i, -1, -1):
                if s[j:i+1] in wordSet and mem[j]:
                    mem[i+1] = True
        return mem[len(s)]

        



if __name__ == '__main__':
    print(Solution().wordBreak("leetcode", ["leet", "code"]))
    print(Solution().wordBreak("s", ["s"]))